### ISSUE TITLE
findAllDependenciesHelp can get into an infinite loop

### ISSUE THREAD

**Issue by PublicParadise** [rtfeldman/node-elm-compiler#39]
I have a case where I can get findAllDependenciesHelp into an infinite loop.
After placing this statement in `findAllDependenciesHelp` before returning the `Promise`:

```
console.log("node-elm-compiler: findAllDependenciesHelp - " + file);
```

I got this output:

```
node-elm-compiler: findAllDependenciesHelp - /Users/bparadie/git/research/elm/mc_client/src/Coral/Icon.elm
node-elm-compiler: findAllDependenciesHelp - /Users/bparadie/git/research/elm/mc_client/src/Coral/Options.elm
node-elm-compiler: findAllDependenciesHelp - /Users/bparadie/git/research/elm/mc_client/src/Coral/Options/Internal.elm
node-elm-compiler: findAllDependenciesHelp - /Users/bparadie/git/research/elm/mc_client/src/Coral/Options.elm
node-elm-compiler: findAllDependenciesHelp - /Users/bparadie/git/research/elm/mc_client/src/Coral/Options/Internal.elm
node-elm-compiler: findAllDependenciesHelp - /Users/bparadie/git/research/elm/mc_client/src/Coral/Options/Internal.elm
node-elm-compiler: findAllDependenciesHelp - /Users/bparadie/git/research/elm/mc_client/src/Coral/Tooltip.elm
node-elm-compiler: findAllDependenciesHelp - /Users/bparadie/git/research/elm/mc_client/src/Coral/Icon.elm
...
```

I looked at the code and I couldn't find any code that would check whether `file` was already in `knownDependencies`. The problem is actually fixed when you add this code to `findAllDependenciesHelp`:

```
if( knownDependencies.indexOf( file ) >= 0 )
{
  return;    
}
```

Now, the interesting thing is that this infinite loop is most likely responsible for a bunch of problems that appear to be `webpack` out-of-memory bugs like [this one here](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/elm-dev/gRnNsc6Oy1A/Qq4HeRasDAAJ). I suspect that `findAllDependenciesHelp` is happily looping and slowly draining V8's memory. 

I have a hunch that this patch will also fix https://github.com/rtfeldman/elm-webpack-loader/issues/62 .

**Comment by PublicParadise** [comment]
Ok, after studying the files involved involved in that infinite loop I am beginning to suspect that `import` statements in comments are interpreted as "real" `import` statements. The code in `findAllDependenciesHelp` doesn't seem to be smart enough to detect import statements in comments: 
```Javascript
/^import\s+([^\s]+)/
```
That means that my proposed fix would prevent the infinite loop but you'd probably get some bogus imports. Not perfect, but IMHO shaking the elm AST would be overkill...

**Comment by chaughawout** [comment]
Might adjust the regex to use a negative lookahead to filter out cases where the import might be buried in a multi-line comment:

`/(?!(.*\s*-}))^import\s+([^\s]+)/`

**Comment by rtfeldman** [comment]
Seems like the best solution might be to write a small parser, one that only detects import statements and comments.

If anyone knows how to write such a (performant) parser in JS, help would be very much appreciated!

**Comment by eeue56** [comment]
published as 4.0.0 in elm-webpack-loader, this now does things in a better way.
